### PR TITLE
Allow to pin a specific runtime version

### PR DESCRIPTION
**Pull request by schellap**
@anurse, @piotrpMSFT, @gkhanna79, @leecow, @richlander

Provide a way to pin the runtime version.

### CODE REVIEW

**Comment by anurse** [comment]
Looks correct to me. /cc @DamianEdwards

**Comment by davidfowl** [comment]
Not sure about this. A cleaner way to handle it would be to support version ranges. 

Min version:

```
"framework": {
    "name": "Microsoft.DotNetCore",
    "version": "1.0.1"
}
```

Exact version:

```
"framework": {
    "name": "Microsoft.DotNetCore",
    "version": "[1.0.1]"
}
```

Future (not for v1.1)

```
"framework": {
    "name": "Microsoft.DotNetCore",
    "version": "1.0.*"
}
```

Wildcard syntax.

**Comment by anurse** [comment]
That seems OK to me, as long as we keep the version range syntax clean and simple (the NuGet syntax is pretty simple: https://docs.nuget.org/create/versioning).

File: Documentation/specs/runtime-configuration-file.md [hunk 1]
```diff
@@ -118,7 +119,10 @@ This section is copied verbatim from an identical section in the input `project.
...
 * `gcConcurrent` - Boolean indicating if background garbage collection should be used (Default: _TBD_). Note: This is designed to mirror the existing [app.config](https://msdn.microsoft.com/en-us/library/yhwwzef8.aspx) setting).
-* `framework` - Indicates the name and version of the shared framework to use when activating the application. The presence of this section indicates that the application is a portable app designed to use a shared redistributable framework.
+* `framework` - Indicates the `name`, `version` and other properties of the shared framework to use when activating the application. The presence of this section indicates that the application is a portable app designed to use a shared redistributable framework.
+  * `autoRollForward` - When `false`, the runtime version is strictly obeyed by the host. When `autoRollForward` is not specified, or specified as `true`, then the runtime from the highest version that differs only in the `SemVer` patch field will be picked up.
```
Comment (gkhanna79) [inline Documentation/specs/runtime-configuration-file.md]: "runtime" - this should be "framework".

File: Documentation/specs/runtime-configuration-file.md [hunk 2]
```diff
@@ -118,7 +119,10 @@ This section is copied verbatim from an identical section in the input `project.
...
+    * For example, if `version=1.0.1` and `autoRollForward` is `true`, the host would load `coreclr.dll` from `1.0.{max}` but will not load from `1.1.0` even if present. When `autoRollForward` is `false`, `coreclr.dll` will be loaded from `1.0.1` strictly.
```
Comment (gkhanna79) [inline Documentation/specs/runtime-configuration-file.md]: There should be no reference to Coreclr.dll. Instead, refer to the "shared framework".

File: Documentation/specs/runtime-configuration-file.md [hunk 3]
```diff
@@ -118,7 +119,10 @@ This section is copied verbatim from an identical section in the input `project.
 
 * `gcServer` - Boolean indicating if the server GC should be used (Default: _TBD_). Note: This is designed to mirror the existing [app.config](https://msdn.microsoft.com/en-us/library/ms229357.aspx) setting)
 * `gcConcurrent` - Boolean indicating if background garbage collection should be used (Default: _TBD_). Note: This is designed to mirror the existing [app.config](https://msdn.microsoft.com/en-us/library/yhwwzef8.aspx) setting).
-* `framework` - Indicates the name and version of the shared framework to use when activating the application. The presence of this section indicates that the application is a portable app designed to use a shared redistributable framework.
+* `framework` - Indicates the `name`, `version` and other properties of the shared framework to use when activating the application. The presence of this section indicates that the application is a portable app designed to use a shared redistributable framework.
```
Comment (GuardRex) [inline Documentation/specs/runtime-configuration-file.md]: Suggestion: Don't imply the presence of the section for standalone apps. You might want to add to the end ...
```
This section will not be present for standalone applications that do not rely upon a shared framework. 
```
Nit: Lists have (for the most part) adopted the comma prior to the conjunction format:
```
the `name`, `version`, and other properties
```

File: Documentation/specs/runtime-configuration-file.md (hunk 1 above)
Comment (GuardRex) [inline Documentation/specs/runtime-configuration-file.md]: Comma splice ...
```
... is not specified, or specified as ...
```
to
```
... is not specified or specified as ...
```
Remove "then," as it is implied that what follows must be true if the conditions are met ...
```
... as `true`, then the runtime ...
```
to
```
... as `true`, the runtime ...
```

**Comment by schellap** [comment]
@davidfowl, that definitely sounds cleaner and I like it.

@gkhanna79 and I discussed and decided that we do not want to pay the cost of building this now. 

However, in ask mode, if allowed, we'll make the necessary changes.

**Comment by anurse** [comment]
We should error out. The `rollforward` property is expected to be a JSON boolean, so that would be a JSON parsing error.

**Pull request merged**